File: samples/clients/sse/sse.py
import requests
import sseclient
import json

base_url = 'http://localhost:8000/api/v1'
username = 'admin'
password = 'wexflow2018'
workflow_id = 41

def login(user, passwd, stay_connected=False):
    url = f'{base_url}/login'
    headers = {'Content-Type': 'application/json'}
    payload = {
        'username': user,
        'password': passwd,
        'stayConnected': stay_connected
    }

    response = requests.post(url, json=payload, headers=headers)
    if response.status_code != 200:
        raise Exception(f'HTTP {response.status_code} - {response.reason}')
    
    return response.json()['access_token']

def start_workflow(token, workflow_id):
    url = f'{base_url}/start?w={workflow_id}'
    headers = {
        'Authorization': f'Bearer {token}'
    }

    response = requests.post(url, headers=headers)
    if response.status_code != 200:
        raise Exception(f'HTTP {response.status_code} - {response.reason}')
    
    return response.json()

def listen_to_sse(url, token):
    headers = {
        'Authorization': f'Bearer {token}',
        'Accept': 'text/event-stream'
    }
    response = requests.get(url, headers=headers, stream=True)

    client = sseclient.SSEClient(response)
    print('SSE connection opened')

    for event in client.events():
        try:
            data = json.loads(event.data)
            print('Received SSE JSON:')
            print(json.dumps(data, indent=2))
            break  # Like JS code, stop after first message
        except Exception as e:
            print(f'Failed to parse SSE JSON: {e}')
            break

    print('SSE connection closed')

# Main execution
try:
    token = login(username, password)
    job_id = start_workflow(token, workflow_id)
    print(f'Workflow {workflow_id} started successfully. Job ID: {job_id}')

    sse_url = f'{base_url}/sse/{workflow_id}/{job_id}'
    listen_to_sse(sse_url, token)

except Exception as e:
    print(f'Error: {e}')

File: samples/clients/sse/dotnet/sse.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text;
using System.IO;

var baseUrl = "http://localhost:8000/api/v1";
var username = "admin";
var password = "wexflow2018";
var workflowId = 41;

using var httpClient = new HttpClient();

async Task<string> LoginAsync(string user, string pass, bool stayConnected = false)
{
  var payload = new
  {
    username = user,
    password = pass,
    stayConnected = stayConnected
  };

  var content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");
  var response = await httpClient.PostAsync($"{baseUrl}/login", content);

  if (!response.IsSuccessStatusCode)
    throw new Exception($"Login failed: HTTP {response.StatusCode} - {response.ReasonPhrase}");

  var json = await response.Content.ReadAsStringAsync();
  var doc = JsonDocument.Parse(json);
  return doc.RootElement.GetProperty("access_token").GetString();
}

async Task<string> StartWorkflowAsync(string token, int workflowId)
{
  var request = new HttpRequestMessage(HttpMethod.Post, $"{baseUrl}/start?w={workflowId}");
  request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

  var response = await httpClient.SendAsync(request);
  if (!response.IsSuccessStatusCode)
    throw new Exception($"Start failed: HTTP {response.StatusCode} - {response.ReasonPhrase}");

  var json = await response.Content.ReadAsStringAsync();
  return JsonSerializer.Deserialize<string>(json);
}

async Task ListenToSseAsync(string url, string token)
{
  var request = new HttpRequestMessage(HttpMethod.Get, url);
  request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
  request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

  var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
  using var stream = await response.Content.ReadAsStreamAsync();
  using var reader = new StreamReader(stream);

  Console.WriteLine("SSE connection opened");

  while (!reader.EndOfStream)
  {
    var line = await reader.ReadLineAsync();
    if (!string.IsNullOrWhiteSpace(line) && line.StartsWith("data: "))
    {
      var json = line.Substring("data: ".Length);
      try
      {
        var doc = JsonDocument.Parse(json);
        Console.WriteLine("Received SSE JSON:");
        Console.WriteLine(doc.RootElement.ToString());
        break; // Close after first message
      }
      catch (Exception ex)
      {
        Console.WriteLine("Failed to parse SSE JSON: " + ex.Message);
      }
    }
  }

  Console.WriteLine("SSE connection closed");
}

try
{
  var token = await LoginAsync(username, password);
  var jobId = await StartWorkflowAsync(token, workflowId);

  Console.WriteLine($"Workflow {workflowId} started. Job ID: {jobId}");

  var sseUrl = $"{baseUrl}/sse/{workflowId}/{jobId}";
  await ListenToSseAsync(sseUrl, token);
}
catch (Exception ex)
{
  Console.WriteLine("Error: " + ex.Message);
}
